src/delta.c

```diff
@@ -38,86 +38,159 @@
 #include "swupd.h"
 #include "xattrs.h"
 
-static void do_delta(char *filename, struct file *file);
-
-void try_delta(struct file *file)
+static bool compute_hash_from_file(char *filename, char *hash)
 {
-	char *filename;
-	struct stat stat;
+	/* TODO: implement this without the indirection of creating a file... */
+	struct file f = {};
 
-	if (file->is_file == 0) {
-		return;
-	}
+	f.filename = filename;
+	f.use_xattrs = true;
 
-	if (file->deltapeer == NULL) {
-		return;
+	populate_file_struct(&f, filename);
+	if (compute_hash(&f, filename) != 0) {
+		return false;
 	}
 
-	if (file->deltapeer->is_file == 0) {
-		return;
-	}
+	hash_assign(&f.hash[0], hash);
+	return true;
+}
 
-	if (file->deltapeer->is_deleted) {
-		return;
+static void apply_one_delta(char *from_file, char *to_staged, char *delta_file, char *to_hash)
+{
+	int ret;
+
+	char *prefixed_from_file;
+	string_or_die(&prefixed_from_file, "%s/%s", path_prefix, from_file);
+
+	ret = apply_bsdiff_delta(prefixed_from_file, to_staged, delta_file);
```

Comment:
Should we check the hash of the prefixed_from_file here?
Better still should we copy the prefixed from file to a secure location and then check the copy?
Otherwise we open ourselves up to the file being changed either before or during the update, which might put bsdiff into an infinite loop.

Reply:
I've kept the behavior of previous delta application code. I think some of these improvements might make sense, but I think they are independent of the main goal here, which is change how we pick what deltas to apply.

> Otherwise we open ourselves up to the file being changed either before or during the update, which might put bsdiff into an infinite loop.

I did not understand how bsdiff could be put in a loop -- but I don't really know its internals. If the a wrong file is given, then it just fail and we fallback to the fullfile.

Reply:
It is only a guess that it can be put into a loop.

Many of these kinds of 'take an input stream and a set of edits to produce an output stream' can have issues, e.g. copy from the last line starting 'A' to here can be affected if you change which that last line happens to be.